api/src/main/java/org/eclipse/microprofile/config/ConfigValueResolver.java

```diff
@@ -0,0 +1,294 @@
+     *         exceptions has been requested using {@link #throwOnFailedConversion()}
+     * @throws java.util.NoSuchElementException if the property isn't present in the configuration and throwing
+     *         exceptions has been requested using {@link #throwOnMissingProperty()}
+     **/
+    <T> T as(Class<T> type, T defaultValue);
+
+    /**
+     * Resolves the property as a simple or array type wrapped in an {@link Optional}.
+     * <p>
+     * If the property is missing, defined empty, the required converter is missing or conversion fails
+     * {@link Optional#empty()} is returned.
+     *
+     * @param type         target type of the conversion, not {@code null}
+     * @return the resolved and converted property value as present or empty {@link Optional}
+     * @throws java.lang.IllegalArgumentException if the property cannot be converted to the specified type and throwing
+     *         exceptions has been requested using {@link #throwOnFailedConversion()}
+     * @throws java.util.NoSuchElementException if the property isn't present in the configuration and throwing
+     *         exceptions has been requested using {@link #throwOnMissingProperty()}
+     **/
+    <T> Optional<T> as(Class<T> type);
+
+    /**
+     * Resolves the property as converted by the provided converter {@link Function}.
+     * <p>
+     * If the property is missing, defined empty or the conversion fails the provided default value is returned.
+     * <p>
+     * This is meant for ad-hoc conversions using lambda expressions as converter to circumvent the need to register a
+     * special {@link org.eclipse.microprofile.config.spi.Converter} for single case use or to allow using different
+     * converter functions for same target type at multiple usages.
+     * <p>
+     * If this method is used in connection with {@link #acceptEmpty()} the empty {@link String} might be passed to the
+     * provided converter function, otherwise the empty string will be considered missing.
+     *
+     * @param type         target type of the conversion, not {@code null}
+     * @param defaultValue any value including null. It is returned in case of missing property or failed conversion
+     *                     unless throwing exceptions has been explicitly requested using
+     *                     {@link #throwOnMissingProperty()} or {@link #throwOnFailedConversion()}.
+     * @return the resolved and converted property value or the default value
+     * @throws java.lang.IllegalArgumentException if the property cannot be converted to the specified type and throwing
+     *         exceptions has been requested using {@link #throwOnFailedConversion()}
+     * @throws java.util.NoSuchElementException if the property isn't present in the configuration and throwing
+     *         exceptions has been requested using {@link #throwOnMissingProperty()}
+     **/
+    <T> T asConvertedBy(Function<String, T> converter, T defaultValue);
```

Comment:
Why are we suddenly using `Function<String,T>` for conversion?

Reply:
This was explained in the PR description. Anyhow, it was discussed 2 weeks ago and I changed it to `Converter`.